lib/elixir/test/elixir/code_normalizer_test.exs

```diff
@@ -0,0 +1,476 @@
+      assert quoted_to_string(quote(do: %Test{} + 1)) == "%Test{} + 1"
+      assert quoted_to_string(quote(do: %Test{foo(1)} + 2)) == "%Test{foo(1)} + 2"
+    end
+
+    test "binary operators" do
+      assert quoted_to_string(quote(do: 1 + 2)) == "1 + 2"
+      assert quoted_to_string(quote(do: [1, 2 | 3])) == "[1, 2 | 3]"
+      assert quoted_to_string(quote(do: [h | t] = [1, 2, 3])) == "[h | t] = [1, 2, 3]"
+      assert quoted_to_string(quote(do: (x ++ y) ++ z)) == "(x ++ y) ++ z"
+      assert quoted_to_string(quote(do: (x +++ y) +++ z)) == "(x +++ y) +++ z"
+    end
+
+    test "unary operators" do
+      assert quoted_to_string(quote(do: not 1)) == "not 1"
+      assert quoted_to_string(quote(do: not foo)) == "not foo"
+      assert quoted_to_string(quote(do: -1)) == "-1"
+      assert quoted_to_string(quote(do: +(+1))) == "+(+1)"
+      assert quoted_to_string(quote(do: !(foo > bar))) == "!(foo > bar)"
+      assert quoted_to_string(quote(do: @foo(bar))) == "@foo bar"
+      assert quoted_to_string(quote(do: identity(&1))) == "identity(&1)"
+    end
+
+    test "access" do
+      assert quoted_to_string(quote(do: a[b])) == "a[b]"
+      assert quoted_to_string(quote(do: a[1 + 2])) == "a[1 + 2]"
+      assert quoted_to_string(quote(do: (a || [a: 1])[:a])) == "(a || [a: 1])[:a]"
+      assert quoted_to_string(quote(do: Map.put(%{}, :a, 1)[:a])) == "Map.put(%{}, :a, 1)[:a]"
+    end
+
+    test "keyword list" do
+      assert quoted_to_string(quote(do: [a: a, b: b])) == "[a: a, b: b]"
+      assert quoted_to_string(quote(do: [a: 1, b: 1 + 2])) == "[a: 1, b: 1 + 2]"
+      assert quoted_to_string(quote(do: ["a.b": 1, c: 1 + 2])) == "[\"a.b\": 1, c: 1 + 2]"
+    end
+
+    test "interpolation" do
+      assert quoted_to_string(quote(do: "foo#{bar}baz")) == ~S["foo#{bar}baz"]
+    end
+
+    test "bit syntax" do
+      ast = quote(do: <<1::8*4>>)
+      assert quoted_to_string(ast) == "<<1::8*4>>"
+
+      ast = quote(do: @type(foo :: <<_::8, _::_*4>>))
+      assert quoted_to_string(ast) == "@type foo :: <<_::8, _::_*4>>"
+
+      ast = quote(do: <<69 - 4::bits-size(8 - 4)-unit(1), 65>>)
+      assert quoted_to_string(ast) == "<<69 - 4::bits-size(8 - 4)-unit(1), 65>>"
+
+      ast = quote(do: <<(<<65>>), 65>>)
+      assert quoted_to_string(ast) == "<<(<<65>>), 65>>"
+
+      ast = quote(do: <<65, (<<65>>)>>)
+      assert quoted_to_string(ast) == "<<65, (<<65>>)>>"
+
+      ast = quote(do: for(<<(a::4 <- <<1, 2>>)>>, do: a))
+      assert quoted_to_string(ast) == "for <<(a::4 <- <<1, 2>>)>> do\n  a\nend"
+    end
+
+    test "charlist" do
+      assert quoted_to_string(quote(do: [])) == "[]"
+      assert quoted_to_string(quote(do: 'abc')) == "'abc'"
+    end
+
+    test "string" do
+      assert quoted_to_string(quote(do: "")) == ~S/""/
+      assert quoted_to_string(quote(do: "abc")) == ~S/"abc"/
+      assert quoted_to_string(quote(do: "#{"abc"}")) == ~S/"#{"abc"}"/
```

Comment:
We need to add some tests on this. The issue is that the formatter works on *unescaped* strings. So if the formatter receives "null byte\t\x00", it will keep those as is. However, if we are working with regular quoted expressions, then that will have the literal bytes, and we need to convert them to the escaped representation. This also need to happen to charlists, atoms, and keyword keys.

Furthermore, if the goal of your changes is to make source modification easy, then we need to:

1. Document the `unescape` flag in Code.string_to_quoted

2. `Code.quoted_to_algebra` needs a flag that says if the strings are unescaped or not (either a :unescaped flag that defaults to false or a :escape flag that defaults to true)

Reply:
Oh, one last one, the Macro.to_string code probably already has the necessary logic to escape string/atom/charlist characters.

Reply:
> 2. Code.quoted_to_algebra needs a flag that says if the strings are unescaped or not (either a :unescaped flag that defaults to false or a :escape flag that defaults to true)

Why is that needed? As it is now, just passing `unescape: false` is enough to preserve the original strings. I'm porting some of the formatter tests to assert that both `Code.format_string!` and `Code.string_to_quoted_with_comments` + `Code.quoted_to_algebra` produce the same output, and it seems to work as expected.

Reply:
If we want to say Code.quoted_to_algebra works with any quoted expression, then it needs to work with any quoted expression, not only unescaped ones. Otherwise it is really some_very_specific_quoted_to_algebra.